Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import hashlib
+import json
+
+from webkitcorepy import string_utils
+from webkitflaskpy import Database
+
+
+class IPCManager(object):
+     DEFAULT_EXPIRATION = 10 * 60
+     DEFAULT_TIMEOUT = 60
+     RETRY = 3
+
+     def __init__(self, *databases):
+         self.databases = databases or [Database()]
+         self.callbacks = dict()
+
+     def register(self, name, callback):
+         if not isinstance(name, str):
+             raise ValueError('Name of registered callback must be a string')
+         if not callable(callback):
+             raise ValueError('Callback must be callable')
+         self.callbacks[name] = callback
+
+     def enqueue(self, name, *args, **kwargs):
+         expiration = kwargs.pop('ex', self.DEFAULT_EXPIRATION)
+         retry = kwargs.pop('retry', self.RETRY)
+         if name not in self.callbacks:
+             raise ValueError("'{}' is not a registered callback".format(name))
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
```

Comment:
Also, should we use `json.dumps` with `sort_keys=True`, also worth adding a unit test for this.
```
def foo(**kwargs):
     print(json.dumps(kwargs))
 
foo(a=1, b=2) # {"a": 1, "b": 2}
foo(b=2, a=1) # {"b": 2, "a": 1}
```

Also, what's the expected behavior if I enqueue same request twice or merge the request?

Reply:
- Yes, arguments would need to be json serializable. In practice, that's an easy requirement to meet because we use this class when receiving an upload, our arguments are essentially always strings.

- Good call on the `sort_keys=True`, will make that change.

- Expected behavior is that we would end up with a single request. Consider, for example, the results database. If we get two identical uploads, we only need to process those once. Likewise, if we get two of the exact same hook from GitHub at the same time, we only want to process one of them.

Reply:
What if one of those two identical uploads arrived after the first one is dequeued? If the callback throw an exception, we will hit the retry logic until it exhausted retries. Or the callback can handle this gracefully.

Or we just use this nice dedupe as an optimization to avoid same request enqueued twice if it's currently in the queue.

Reply:
If two identical uploads arrive after the first one is dequeued, we have 2 basic options:
- The dequeue completed
- The dequeue triggered a retry
In the first case, we'll re-queue the same task and repeat the operation. Note that case 1 covers the task completing successfully, the task being somehow invalid (like the named task doesn't exist on the worker) and the task exhausting its retry.

In the second case, the incoming task will overwrite the retry count, basically reseting it.

There is also the question of what happens if something arrives _during_ the dequeue, which looks very much like the case where two identical tasks are collapsed into one.